pkg/preparation/sqlrepo/shards.go

```diff
@@ -0,0 +1,176 @@
+package sqlrepo
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ipfs/go-cid"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/multiformats/go-varint"
+	configmodel "github.com/storacha/guppy/pkg/preparation/configurations/model"
+	"github.com/storacha/guppy/pkg/preparation/shards"
+	"github.com/storacha/guppy/pkg/preparation/shards/model"
+	"github.com/storacha/guppy/pkg/preparation/sqlrepo/util"
+	"github.com/storacha/guppy/pkg/preparation/types/id"
+)
+
+// Byte length of a CBOR encoded CAR header with zero roots.
+const noRootsHeaderLen = 17
+
+var _ shards.Repo = (*repo)(nil)
+
+func (r *repo) CreateShard(ctx context.Context, uploadID id.UploadID) (*model.Shard, error) {
+	shard, err := model.NewShard(uploadID)
+	if err != nil {
+		return nil, err
+	}
+
+	err = model.WriteShardToDatabase(shard, func(id id.ShardID, uploadID id.UploadID, cid *cid.Cid, state model.ShardState) error {
+		_, err := r.db.ExecContext(ctx, `
+			INSERT INTO shards (
+				id,
+				upload_id,
+				cid,
+				state
+			) VALUES (?, ?, ?, ?)`,
+			id,
+			uploadID,
+			Null(cid),
+			state,
+		)
+		return err
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to write shard for upload %s: %w", uploadID, err)
+	}
+
+	return shard, nil
+}
+
+func (r *repo) ShardsForUploadByStatus(ctx context.Context, uploadID id.UploadID, state model.ShardState) ([]*model.Shard, error) {
+	rows, err := r.db.QueryContext(ctx, `
+		SELECT
+			id,
+			upload_id,
+			cid,
+			state
+		FROM shards
+		WHERE upload_id = ?
+		  AND state = ?`,
+		uploadID,
+		state,
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var shards []*model.Shard
+	for rows.Next() {
+		shard, err := model.ReadShardFromDatabase(func(
+			id *id.ShardID,
+			uploadID *id.UploadID,
+			cid **cid.Cid,
+			state *model.ShardState,
+		) error {
+			return rows.Scan(id, uploadID, cid, state)
+		})
+		if err != nil {
+			return nil, err
+		}
+		if shard == nil {
+			continue
+		}
+		shards = append(shards, shard)
+	}
+	return shards, nil
+}
+
+func (r *repo) AddNodeToShard(ctx context.Context, shardID id.ShardID, nodeCID cid.Cid) error {
+	_, err := r.db.ExecContext(ctx, `
+		INSERT INTO nodes_in_shards (
+			node_cid,
+			shard_id
+		) VALUES (?, ?)`,
+		nodeCID.Bytes(),
+		shardID,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to add node %s to shard %s: %w", nodeCID, shardID, err)
+	}
+	return nil
+}
+
+func (r *repo) RoomInShard(ctx context.Context, shard *model.Shard, nodeCID cid.Cid, config *configmodel.Configuration) (bool, error) {
```

Comment:
I think this is the only method that I kind of think there should be a refactor on.

You're putting domain logic (how shards get sized) into the DB repo interface.

I would rather see a method like:

```golang

```golang

type ShardNode interface {
   Cid() cid.Cid
   Size() uint64
}

func (r *Repo) IterateShardNodes(ctx context.Context, shard *model.Shard, cb func(ShardNode) error) error {
...
}
```

and then the rest of this logic (i.e. RoomInShard calculation) goes up to the Shards module.

I think this is my only blocker so far.